src/gt4sd/frameworks/enzeptional/processing.py

```diff
@@ -21,237 +21,604 @@
 # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 # SOFTWARE.
 #
-"""enzeptional - data processing utilities."""
-
 from abc import ABC
-from typing import Generic, List, Optional, Tuple, TypeVar, Union
-
-import numpy as np
 import torch
+import numpy as np
+from typing import Any, Dict, List, Optional, Tuple, Union
 from tape.datasets import pad_sequences
 from tape.registry import registry
 from tape.tokenizers import TAPETokenizer
-from transformers import AutoModelWithLMHead, AutoTokenizer
+from transformers import (
+    AutoModel,
+    EsmForMaskedLM,
+    AutoTokenizer,
+    T5Tokenizer,
+)
+import math
+import random
+import logging
+from itertools import product as iter_product
 
-from ..torch import device_claim
 
-T = TypeVar("T")  # used for sample embedding
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
 
+# os.environ["TRANSFORMERS_CACHE"] = "~/.cache/huggingface/"
+# torch.hub.set_dir("/dccstor/yna/.cache/torch/hub")
 
-class Embedding(ABC, Generic[T]):
-    """Abstract embedding class."""
 
-    def embed_one(self, sample: T) -> np.ndarray:
-        """Embed one sample.
+class ModelCache:
+    """
+    A simple cache mechanism for storing and retrieving models.
+    """
+
+    def __init__(self):
+        """
+        Initializes the cache as an empty dictionary.
+        """
+        self.cache = {}
+
+    def get(self, key):
+        """
+        Retrieves a model from the cache using the given key.
 
         Args:
-            sample: sample representation.
+            key: The key used to store the model.
 
         Returns:
-            embedding vector for the sample.
+            The model associated with the key, or None if not found.
         """
-        return self.__call__([sample])
+        return self.cache.get(key)
 
-    def __call__(self, samples: List[T]) -> np.ndarray:
-        """Embed multiple samples sample.
+    def add(self, key, model):
+        """
+        Adds a model to the cache with the specified key.
 
         Args:
-            samples: a list of sample representations.
+            key: The key to associate with the model.
+            model: The model to be cached.
+        """
+        self.cache[key] = model
 
-        Returns:
-            embedding vectors for the samples.
+
+model_cache = ModelCache()
+
+
+def get_device(device: Optional[Union[torch.device, str]] = None) -> torch.device:
```

Comment:
do not rewrite code that already exists, simply import it: https://github.com/GT4SD/gt4sd-core/blob/94f009132968ac4558e8256a4afcbf740065c5cd/src/gt4sd/frameworks/torch/__init__.py#L65